### PR TITLE
[ch46954] Refactor for 1.0 release

### PR DESCRIPTION
**Pull request by alexcombessie**
Clubhouse story [ch46954](https://app.clubhouse.io/dataiku/story/46954/ux-enhancement-and-backend-refactoring)

### CODE REVIEW

**Comment by alexcombessie** [comment]
I need to use the REST API instead of the Python client (which really s*** a**) https://docs.microsoft.com/en-gb/azure/cognitive-services/text-analytics/quickstarts/python
https://westcentralus.dev.cognitive.microsoft.com/docs/services/TextAnalytics-v2-1/operations/5ac4251d5b4ccd1554da7634

**Comment by alexcombessie** [comment]
Clubhouse story [ch46954](https://app.clubhouse.io/dataiku/story/46954/ux-enhancement-and-backend-refactoring)

**Comment by alexcombessie** [comment]
👋 @MayeulRousselet

**Comment by alexcombessie** [comment]
@MayeulRousselet ready for review for API v3.0. I tested thorougly on http://dku17.dataiku.com:30800/projects/AZURENLP/scenarios/TESTAZURENLP/settings

**Comment by alexcombessie** [comment]
Gentle followup @MayeulRousselet for the review 🥺

**Comment by alexcombessie** [comment]
@StanislasGuinel thanks for accepting the review request 🏅

File: custom-recipes/azure-cognitive-services-nlp-key-phrases-extraction/recipe.py [hunk 1]
```diff
@@ -1,62 +1,98 @@
...
+input_dataset_name = get_input_names_for_role("input_dataset")[0]
 input_dataset = dataiku.Dataset(input_dataset_name)
 input_schema = input_dataset.read_schema()
-input_columns_names = [col['name'] for col in input_schema]
-keyphrases_column = generate_unique('keyphrases', input_columns_names)
+input_columns_names = [col["name"] for col in input_schema]
 
-output_dataset_name = get_output_names_for_role('output_dataset')[0]
+output_dataset_name = get_output_names_for_role("output_dataset")[0]
 output_dataset = dataiku.Dataset(output_dataset_name)
```
Comment (StanislasGuinel) [inline custom-recipes/azure-cognitive-services-nlp-key-phrases-extraction/recipe.py:38]: Maybe you could also validate the language_column
```suggestion
if text_language == "language_column":
    validate_column_input(language_column, input_columns_names)
```

File: custom-recipes/azure-cognitive-services-nlp-key-phrases-extraction/recipe.py [hunk 2]
```diff
@@ -1,62 +1,98 @@
...
 
-if text_column is None or len(text_column) == 0:
-    raise ValueError("You must specify the input text column")
-if text_column not in input_columns_names:
-    raise ValueError("Column '{}' is not present in the input dataset".format(text_column))
+validate_column_input(text_column, input_columns_names)
```
Comment (StanislasGuinel) [inline custom-recipes/azure-cognitive-services-nlp-key-phrases-extraction/recipe.py:39]: Maybe you could also validate the language_column
```suggestion
validate_column_input(text_column, input_columns_names)
if text_language == "language_column":
    validate_column_input(language_column, input_columns_names)
```

File: python-lib/azure_nlp_api_client.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+# ==============================================================================
+
+API_EXCEPTIONS = requests.RequestException
+
+# ==============================================================================
+# CLASS AND FUNCTION DEFINITION
+# ==============================================================================
+
+
+class get_client:
```
[the remark this answers is not in the document]
Comment (alexcombessie) [reply python-lib/azure_nlp_api_client.py:19]: Indeed, forgot this one! Thanks, I changed it.

File: custom-recipes/azure-cognitive-services-nlp-key-phrases-extraction/recipe.json [hunk 4]
```diff
@@ -21,38 +21,161 @@
...
         {
             "name": "language",
             "label": "Language",
-            "type": "STRING",
-            "description":"in ISO 639-1 format",
-            "mandatory": false
+            "description": "List of supported languages",
+            "type": "SELECT",
+            "mandatory": true,
+            "selectChoices": [
```
[the remark this answers is not in the document]
Comment (alexcombessie) [reply custom-recipes/azure-cognitive-services-nlp-key-phrases-extraction/recipe.json:52]: Actually, each recipe has a slightly different list of languages. This is Azure's weirdness, not mine: https://docs.microsoft.com/en-us/azure/cognitive-services/text-analytics/language-support?tabs=sentiment-analysis

Until Azure harmonize this across their API, I would rather keep it like this.

File: parameter-sets/api-configuration/parameter-set.json [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        },
+        {
+            "name": "azure_region",
+            "label": "Azure region",
+            "description": "If empty, attempts to ascertain endpoint from the environment.",
+            "type": "STRING",
+            "defaultValue": "eastus",
+            "mandatory": false
+        },
+        {
```
[the remark this answers is not in the document]
Comment (alexcombessie) [reply parameter-sets/api-configuration/parameter-set.json:31]: It's a valid point. In all other NLP API plugins, I have kept these parameters at top level. My rationale is that:
- Parameter set configuration is for admins, not regular users. I assume admins can handle this additional complexity.
- The Quota parameters definitely need tuning, according to each customer number of users and pricing plan with the API provider. Hence I don't want to "hide" them in an Advanced options, as I want them to be easy to find.

File: custom-recipes/azure-cognitive-services-nlp-key-phrases-extraction/recipe.py (hunk 2 above)
Comment (alexcombessie) [reply custom-recipes/azure-cognitive-services-nlp-key-phrases-extraction/recipe.py:39]: Good point, added to all recipes (exc. language detection, which doesn't have the param)

File: python-lib/azure_nlp_api_client.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    def __init__(self, api_configuration_preset):
+        if api_configuration_preset is None or api_configuration_preset == {}:
```
[the remark this answers is not in the document]
Comment (alexcombessie) [reply python-lib/azure_nlp_api_client.py:21]: Good point, much more elegant

File: python-lib/azure_nlp_api_client.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            raise ValueError("No Azure credentials provided, please enter an API configuration preset")
+        self.api_key = str(api_configuration_preset.get("azure_api_key", ""))
+        self.region = str(api_configuration_preset.get("azure_region", ""))
+        self.endpoint = "https://{}.api.cognitive.microsoft.com".format(self.region)
+        if self.api_key is None or self.api_key == "":
```
[the remark this answers is not in the document]
Comment (alexcombessie) [reply python-lib/azure_nlp_api_client.py:26]: Good point, much more elegant

**Review by StanislasGuinel — APPROVED** [review]
All is good !!

**Pull request merged**